src/extension.ts

```diff
@@ -1,354 +1,14 @@
 import * as vscode from 'vscode';
-import {
-  ConnectionOption,
-  ExtendedMap,
-  createConnectStatusBarItem,
-  createConnectTextAndTooltip,
-  createConnectionOptions,
-  createConnectionQuickPick,
-  getTempDir,
-  Logger,
-  Toaster,
-} from './util';
-import { DhcService, RunCommandCodeLensProvider } from './services';
-import { DhServiceRegistry } from './services';
-import {
-  DOWNLOAD_LOGS_CMD,
-  RUN_CODE_COMMAND,
-  RUN_SELECTION_COMMAND,
-  SELECT_CONNECTION_COMMAND,
-} from './common';
-import { OutputChannelWithHistory } from './util/OutputChannelWithHistory';
+import { ExtensionController } from './services';
 
-const logger = new Logger('extension');
+let controller: ExtensionController | null = null;
 
-export function activate(context: vscode.ExtensionContext) {
-  let selectedConnectionUrl: string | null = null;
-  let selectedDhService: DhcService | null = null;
+export function activate(context: vscode.ExtensionContext): void {
+  const controller = new ExtensionController(context);
```

Comment:
You're not assigning to the top-level controller, it seems that's never assigned to? Also not sure you need to... but should you check if controller is `null` first and throw if it's not?
Is it even necessary to keep track of the `controller`? You're just using it to call `dispose` in the `deactivate`, but if I'm reading the VSCode API docs correctly, the subscriptions `dispose` should already be called when the extension is deactivated: https://code.visualstudio.com/api/references/vscode-api#ExtensionContext

Reply:
Fixed